# Tucil 1 IF3270 Machine Learning dengan Python

### 13519153 Maximillian Lukman - 13519162 Irvin Andryan Pratomo

In [1]:
from sklearn import datasets, tree, model_selection, preprocessing, metrics, svm, cluster, linear_model, neural_network

In [2]:
import six
import sys
sys.modules['sklearn.externals.six'] = six

In [3]:
from id3 import Id3Estimator, export_text

In [4]:
import pandas as pd

In [5]:
#load data from sklearn.datasets
cancer = datasets.load_breast_cancer()

In [6]:
#load data from csv file
tennis = pd.read_csv('play_tennis.csv')

In [7]:
#transform label for tennis
le = preprocessing.LabelEncoder()

tennis = tennis.drop('day', axis = 1) #----> day perlu dimasukkan sebagai data training gak ya?
# tennis['day'] = le.fit_transform(tennis['day']) #kalau day gamasuk line ini hapus
tennis['outlook'] = le.fit_transform(tennis['outlook'])
tennis['temp'] = le.fit_transform(tennis['temp'])
tennis['humidity'] = le.fit_transform(tennis['humidity'])
tennis['wind'] = le.fit_transform(tennis['wind'])
tennis['play'] = le.fit_transform(tennis['play'])

In [8]:
X_cancer = cancer.data
y_cancer = cancer.target

In [9]:
X_tennis = tennis.drop('play', axis = 1)
y_tennis = tennis.play

In [10]:
#Set training data to 80% and test data to 20% for cancer
X_train_cancer, X_test_cancer, y_train_cancer, y_test_cancer = model_selection.train_test_split(X_cancer,y_cancer,test_size=0.2,shuffle=True)

In [11]:
#Set training data to 80% and test data to 20% for tennis
X_train_tennis, X_test_tennis, y_train_tennis, y_test_tennis = model_selection.train_test_split(X_tennis,y_tennis,test_size=0.2,shuffle=True)

## Decision Tree Classifier

In [12]:
decision_tree = tree.DecisionTreeClassifier(criterion='entropy', max_depth=20)

In [13]:
#decision tree for tennis
decision_tree_tennis = decision_tree.fit(X_train_tennis, y_train_tennis)
r_tennis = tree.export_text(decision_tree_tennis)
print(r_tennis)

|--- feature_0 <= 0.50
|   |--- class: 1
|--- feature_0 >  0.50
|   |--- feature_3 <= 0.50
|   |   |--- feature_1 <= 1.50
|   |   |   |--- class: 0
|   |   |--- feature_1 >  1.50
|   |   |   |--- class: 1
|   |--- feature_3 >  0.50
|   |   |--- feature_0 <= 1.50
|   |   |   |--- class: 1
|   |   |--- feature_0 >  1.50
|   |   |   |--- feature_1 <= 1.00
|   |   |   |   |--- class: 1
|   |   |   |--- feature_1 >  1.00
|   |   |   |   |--- class: 0



In [14]:
prediction_tennis_decisionTreeClassifier = decision_tree_tennis.predict(X_test_tennis)
print("Decision Tree Classifier: Tennis")
print("Accuracy:", metrics.accuracy_score(y_test_tennis, prediction_tennis_decisionTreeClassifier))
print("F1 Macro avg:", metrics.f1_score(y_test_tennis, prediction_tennis_decisionTreeClassifier, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_tennis, prediction_tennis_decisionTreeClassifier, average="weighted"))

Decision Tree Classifier: Tennis
Accuracy: 0.3333333333333333
F1 Macro avg: 0.25
F1 Weighted avg: 0.16666666666666666


In [15]:
#decision tree for cancer
decision_tree_cancer = decision_tree.fit(X_train_cancer, y_train_cancer)
r_cancer = tree.export_text(decision_tree_cancer)
print(r_cancer)

|--- feature_27 <= 0.14
|   |--- feature_20 <= 16.83
|   |   |--- feature_13 <= 38.60
|   |   |   |--- feature_21 <= 30.15
|   |   |   |   |--- class: 1
|   |   |   |--- feature_21 >  30.15
|   |   |   |   |--- feature_14 <= 0.00
|   |   |   |   |   |--- class: 0
|   |   |   |   |--- feature_14 >  0.00
|   |   |   |   |   |--- feature_24 <= 0.14
|   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- feature_24 >  0.14
|   |   |   |   |   |   |--- feature_22 <= 89.08
|   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |--- feature_22 >  89.08
|   |   |   |   |   |   |   |--- class: 0
|   |   |--- feature_13 >  38.60
|   |   |   |--- feature_15 <= 0.01
|   |   |   |   |--- class: 0
|   |   |   |--- feature_15 >  0.01
|   |   |   |   |--- feature_13 <= 39.15
|   |   |   |   |   |--- class: 0
|   |   |   |   |--- feature_13 >  39.15
|   |   |   |   |   |--- class: 1
|   |--- feature_20 >  16.83
|   |   |--- feature_21 <= 19.91
|   |   |   |--- class: 1
|   |   |--- fe

In [16]:
prediction_cancer_decisionTreeClassifier = decision_tree_cancer.predict(X_test_cancer)
print("Decision Tree Classifier: Cancer")
print("Accuracy:", metrics.accuracy_score(y_test_cancer, prediction_cancer_decisionTreeClassifier))
print("F1 Macro avg:", metrics.f1_score(y_test_cancer, prediction_cancer_decisionTreeClassifier, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_cancer, prediction_cancer_decisionTreeClassifier, average="weighted"))

Decision Tree Classifier: Cancer
Accuracy: 0.9649122807017544
F1 Macro avg: 0.9594017094017093
F1 Weighted avg: 0.9643874643874643


## Id3Estimator

In [17]:
estimator = Id3Estimator()

In [18]:
estimator_tennis = estimator.fit(X_train_tennis, y_train_tennis)

In [19]:
prediction_tennis_Id3 = estimator_tennis.predict(X_test_tennis)
print("Id3Estimator: Tennis")
print("Accuracy:", metrics.accuracy_score(y_test_tennis, prediction_tennis_Id3))
print("F1 Macro avg:", metrics.f1_score(y_test_tennis, prediction_tennis_Id3, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_tennis, prediction_tennis_Id3, average="weighted"))

Id3Estimator: Tennis
Accuracy: 1.0
F1 Macro avg: 1.0
F1 Weighted avg: 1.0


In [20]:
estimator_tennis_tree = export_text(estimator_tennis.tree_)
print(estimator_tennis_tree)


0 <=0.50: 1 (3) 
0 >0.50
|   2 <=0.50
|   |   1 <=1.50: 0 (1) 
|   |   1 >1.50: 0 (1/1) 
|   2 >0.50
|   |   3 <=0.50
|   |   |   1 <=1.00: 0 (1) 
|   |   |   1 >1.00: 1 (1) 
|   |   3 >0.50: 1 (3) 



In [21]:
estimator_cancer = estimator.fit(X_train_cancer, y_train_cancer)

In [22]:
estimator_cancer_tree = export_text(estimator_cancer.tree_, cancer['feature_names'])
print(estimator_cancer_tree)


worst concave points <=0.14
|   worst radius <=16.83
|   |   area error <=38.60
|   |   |   worst texture <=30.15: 1 (216) 
|   |   |   worst texture >30.15
|   |   |   |   texture error <=0.86: 0 (1) 
|   |   |   |   texture error >0.86
|   |   |   |   |   worst smoothness <=0.14: 1 (27) 
|   |   |   |   |   worst smoothness >0.14
|   |   |   |   |   |   mean radius <=12.23: 1 (6) 
|   |   |   |   |   |   mean radius >12.23: 0 (1) 
|   |   area error >38.60
|   |   |   mean compactness <=0.06: 0 (2) 
|   |   |   mean compactness >0.06
|   |   |   |   radius error <=0.42: 0 (1) 
|   |   |   |   radius error >0.42: 1 (9) 
|   worst radius >16.83
|   |   mean texture <=16.19: 1 (6) 
|   |   mean texture >16.19
|   |   |   worst concavity <=0.20
|   |   |   |   mean compactness <=0.07: 0 (3) 
|   |   |   |   mean compactness >0.07: 1 (2) 
|   |   |   worst concavity >0.20: 0 (14) 
worst concave points >0.14
|   worst perimeter <=114.45
|   |   mean texture <=20.30
|   |   |   area error 

In [23]:
prediction_cancer_Id3 = estimator_cancer.predict(X_test_cancer)
print("Id3Estimator: Cancer")
print("Accuracy:", metrics.accuracy_score(y_test_cancer, prediction_cancer_Id3))
print("F1 Macro avg:", metrics.f1_score(y_test_cancer, prediction_cancer_Id3, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_cancer, prediction_cancer_Id3, average="weighted"))

Id3Estimator: Cancer
Accuracy: 0.9649122807017544
F1 Macro avg: 0.9599859599859599
F1 Weighted avg: 0.9646659646659645


## K-Means

In [24]:
kmeans = cluster.KMeans(n_clusters=2)

In [25]:
kmeans_tennis = kmeans.fit(X_train_tennis,y_train_tennis)

In [26]:
kmeans_tennis_centroid = pd.DataFrame(kmeans_tennis.cluster_centers_.transpose())
kmeans_tennis_centroid.index = tennis.columns[:-1].tolist()
kmeans_tennis_centroid.columns = ["Centroid 1", "Centroid 2"]
print(kmeans_tennis_centroid)

          Centroid 1  Centroid 2
outlook     0.666667         1.6
temp        0.333333         1.8
humidity    0.833333         0.4
wind        0.666667         0.6


In [27]:
prediction_tennis_kmeans = kmeans_tennis.predict(X_test_tennis)
print("K-Means: Tennis")
print("Accuracy:", metrics.accuracy_score(y_test_tennis, prediction_tennis_kmeans))
print("F1 Macro avg:", metrics.f1_score(y_test_tennis, prediction_tennis_kmeans, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_tennis, prediction_tennis_kmeans, average="weighted"))

K-Means: Tennis
Accuracy: 0.3333333333333333
F1 Macro avg: 0.25
F1 Weighted avg: 0.16666666666666666


In [28]:
kmeans_cancer = kmeans.fit(X_train_cancer,y_train_cancer)

In [29]:
kmeans_cancer_centroid = pd.DataFrame(kmeans_cancer.cluster_centers_.transpose())
kmeans_cancer_centroid.index = cancer['feature_names'].tolist()
kmeans_cancer_centroid.columns = ["Centroid 1", "Centroid 2"]
print(kmeans_cancer_centroid)

                         Centroid 1   Centroid 2
mean radius               12.610020    19.298868
mean texture              18.700544    21.845377
mean perimeter            81.506275   127.647170
mean area                500.685673  1176.164151
mean smoothness            0.094454     0.101690
mean compactness           0.091569     0.147190
mean concavity             0.064158     0.175125
mean concave points        0.033647     0.100182
mean symmetry              0.178926     0.189862
mean fractal dimension     0.063479     0.060556
radius error               0.303444     0.756993
texture error              1.227403     1.232891
perimeter error            2.146616     5.378491
area error                23.930883    98.083113
smoothness error           0.007090     0.006636
compactness error          0.023894     0.031703
concavity error            0.029568     0.042593
concave points error       0.010599     0.015774
symmetry error             0.020449     0.020377
fractal dimension er

In [30]:
prediction_cancer_kmeans = kmeans_cancer.predict(X_test_cancer)
print("K-Means: Cancer")
print("Accuracy:", metrics.accuracy_score(y_test_cancer, prediction_cancer_kmeans))
print("F1 Macro avg:", metrics.f1_score(y_test_cancer, prediction_cancer_kmeans, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_cancer, prediction_cancer_kmeans, average="weighted"))

K-Means: Cancer
Accuracy: 0.11403508771929824
F1 Macro avg: 0.10236220472440946
F1 Weighted avg: 0.06824146981627298


## Logistic Regression

In [31]:
lr = linear_model.LogisticRegression(class_weight='balanced', max_iter=2500, random_state=0)

In [32]:
lr_tennis = lr.fit(X_train_tennis,y_train_tennis)

In [33]:
prediction_tennis_lr = lr_tennis.predict(X_test_tennis)
print("Logistic Regression: Tennis")
print("Accuracy:", metrics.accuracy_score(y_test_tennis, prediction_tennis_lr))
print("F1 Macro avg:", metrics.f1_score(y_test_tennis, prediction_tennis_lr, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_tennis, prediction_tennis_lr, average="weighted"))

Logistic Regression: Tennis
Accuracy: 1.0
F1 Macro avg: 1.0
F1 Weighted avg: 1.0


In [34]:
lr_cancer = lr.fit(X_train_cancer,y_train_cancer)

In [35]:
prediction_cancer_lr = lr_cancer.predict(X_test_cancer)
print("Logistic Regression: Cancer")
print("Accuracy:", metrics.accuracy_score(y_test_cancer, prediction_cancer_lr))
print("F1 Macro avg:", metrics.f1_score(y_test_cancer, prediction_cancer_lr, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_cancer, prediction_cancer_lr, average="weighted"))

Logistic Regression: Cancer
Accuracy: 0.9649122807017544
F1 Macro avg: 0.9599859599859599
F1 Weighted avg: 0.9646659646659645


## Neural Network

In [36]:
neural = neural_network.MLPClassifier(max_iter=1000)

In [37]:
neural_tennis = neural.fit(X_train_tennis,y_train_tennis)

In [38]:
prediction_tennis_neural = neural_tennis.predict(X_test_tennis)
print("Neural Network: Tennis")
print("Accuracy:", metrics.accuracy_score(y_test_tennis, prediction_tennis_neural))
print("F1 Macro avg:", metrics.f1_score(y_test_tennis, prediction_tennis_neural, average="macro"))
print("F1 Weighted:", metrics.f1_score(y_test_tennis, prediction_tennis_neural, average="weighted"))

Neural Network: Tennis
Accuracy: 0.6666666666666666
F1 Macro avg: 0.6666666666666666
F1 Weighted: 0.6666666666666666


In [39]:
neural_cancer = neural.fit(X_train_cancer,y_train_cancer)

In [40]:
prediction_cancer_neural = neural_cancer.predict(X_test_cancer)
print("Neural Network: Cancer")
print("Accuracy:", metrics.accuracy_score(y_test_cancer, prediction_cancer_neural))
print("F1 Macro avg:", metrics.f1_score(y_test_cancer, prediction_cancer_neural, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_cancer, prediction_cancer_neural, average="weighted"))

Neural Network: Cancer
Accuracy: 0.9473684210526315
F1 Macro avg: 0.9391025641025641
F1 Weighted avg: 0.9465811965811965


## SVM

In [41]:
svc = svm.SVC()

In [42]:
svc_tennis = svc.fit(X_train_tennis,y_train_tennis)

In [43]:
prediction_tennis_svm = svc_tennis.predict(X_test_tennis)
print("SVM: Tennis")
print("Accuracy:", metrics.accuracy_score(y_test_tennis, prediction_tennis_svm))
print("F1 Macro avg:", metrics.f1_score(y_test_tennis, prediction_tennis_svm, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_tennis, prediction_tennis_svm, average="weighted"))

SVM: Tennis
Accuracy: 0.3333333333333333
F1 Macro avg: 0.25
F1 Weighted avg: 0.16666666666666666


In [44]:
svc_cancer = svc.fit(X_train_cancer,y_train_cancer)

In [45]:
prediction_cancer_svm = svc_cancer.predict(X_test_cancer)
print("SVM: Cancer")
print("Accuracy:", metrics.accuracy_score(y_test_cancer, prediction_cancer_svm))
print("F1 Macro avg:", metrics.f1_score(y_test_cancer, prediction_cancer_svm, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_cancer, prediction_cancer_svm, average="macro"))

SVM: Cancer
Accuracy: 0.9473684210526315
F1 Macro avg: 0.9381555153707053
F1 Weighted avg: 0.9381555153707053


## Analisis 

### Perbandingan Nilai Akurasi dan F1 Score dari masing-masing Algoritma pada dataset Play Tennis

Setelah melakukan pembelajaran dengan beberapa algoritma di atas, didapatkan nilai tertinggi untuk Accuracy dan kedua F1 Score dengan menggunakan algoritma Id3Estimator dan LogisticRegression dengan skor 1. Sedangkan nilai terendah untuk dataset Play Tennis didapat dengan menggunakan Decision Tree, KMeans dan SVM dengan skor 0.333 untuk Accuracy dan 0.25 / 0.167 untuk F1 Macro / Weighted average. Terakhir, untuk pembelajaran Neural Network diperoleh skor 0.667 untuk Accuracy dan kedua F1 Score 

### Perbandingan Nilai Akurasi dan F1 Score dari masing-masing Algoritma pada dataset Breast Cancer

Setelah melakukan pembelajaran dengan beberapa algoritma di atas, didapatkan nilai tertinggi dengan menggunakan Decision Tree, Id3Estimator, dan LogisticRegression dengan 0965 untuk Accuracy, 0.960 untuk F1 Macro avg, dan 0.965 untuk F1 Weighted avg. Sedangkan yang terendah didapat dengan pembelajaran KMeans dengan 0.114 untuk Accuracy, 0.102 untuk F1 Macro avg, dan 0.068 untuk F1 Weighted avg.

### Kesimpulan

Dengan melihat hasil atau nilai yang dihasilkan dari pembelajaran berbagai algoritma, kami melihat pola di mana skor yang dihasilkan untuk dataset breast cancer relatif lebih besar daripada untuk dataset play tennis. Hal ini kami asumsikan dikarenakan jumlah data yang sangat jauh berbeda di mana dataset brease